File: src/lib/__tests__/coverage.test.ts
import { describe, expect, it } from "vitest";
import { isInKorea } from "../coverage";
import seed from "../data/osm-walk-nodes.json";
import cases from "./fixtures/korea-boundary-cases.json";

describe("isInKorea — 국경 폴리곤 판정", () => {
  it("공유 golden 전수를 맞힌다", () => {
    for (const c of cases.cases) {
      expect(isInKorea(c.lat, c.lng), c.name).toBe(c.inside);
    }
  });

  it("사각형 밖 좌표는 여전히 밖이다", () => {
    expect(isInKorea(37.7749, -122.4194)).toBe(false); // 샌프란시스코
    expect(isInKorea(0, 0)).toBe(false); // 널 아일랜드
  });

  it("OSM seed 노드 전수가 폴리곤 안이다", () => {
    // 빌드 스크립트의 insideRings와 coverage.ts의 ray casting은 서로 다른 코드다.
    // 이 단언이 두 구현의 합의를 8만 점으로 확인한다(spec §3 방어 2) — 링이 잘못
    // 실렸거나 두 구현이 어긋나면 "국내인데 제공 지역 밖"이라는 조용한 결함이
    // 되는데, 그 방향은 종전 사각형에는 없던 새 실패 모드다.
    const nodes = (seed as unknown as { nodes: Array<[number, number, number, number]> }).nodes;
    expect(nodes.length).toBeGreaterThan(70_000);
    const stray = nodes.find((n) => !isInKorea(n[1], n[2]));
    expect(stray && `노드 ${stray[0]}(${stray[1]}, ${stray[2]})`).toBeUndefined();
  });
});

File: android/kit/src/test/kotlin/space/dodoplanet/gildongmu/kit/CoverageTest.kt
package space.dodoplanet.gildongmu.kit

import kotlinx.serialization.Serializable
import kotlin.test.Test
import kotlin.test.assertContentEquals
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class CoverageTest {
    @Test fun `한국 좌표는 커버리지 안`() {
        assertTrue(isInKorea(37.5665, 126.978))
        assertFalse(isInKorea(37.7749, -122.4194))
    }

    /** 프리필터를 상수(≤132.0)로 두었을 때 잘려 나가던 구간 — 독도 영해 링은 동경 132.12까지 뻗는다. */
    @Test fun `프리필터가 링을 잘라내지 않는다`() {
        assertTrue(isInKorea(37.24, 132.05))
        assertFalse(isInKorea(37.24, 132.2))
    }

    @Serializable
    private data class BoundaryCaseFile(val cases: List<Case>) {
        @Serializable
        data class Case(val name: String, val lat: Double, val lng: Double, val inside: Boolean)
    }

    /** 웹 `coverage.test.ts`·Kit `CoverageTests`와 같은 공유 fixture — 국경 판정 드리프트 가드. */
    @Test fun `국경 공유 golden 전수`() {
        val cases = Fixtures.sharedJson("korea-boundary-cases.json", BoundaryCaseFile.serializer()).cases
        assertTrue(cases.size >= 15)
        // inside: true 케이스가 아홉이라, 리소스를 못 읽어 링이 비면(전 좌표 "밖") 여기서 즉시 실패한다.
        for (c in cases) assertEquals(c.inside, isInKorea(c.lat, c.lng), c.name)
    }

    /** 리소스는 웹 정본의 바이트 동일 사본이다(iOS `korea-boundary-drift.test.ts`와 같은 계약). */
    @Test fun `국경 링 리소스는 웹 정본과 바이트 동일`() {
        val web = Fixtures.repoRoot.resolve("src/lib/data/korea-boundary.json").readBytes()
        val resource = checkNotNull(javaClass.getResourceAsStream("/korea-boundary.json")) { "리소스 부재" }.use { it.readBytes() }
        assertContentEquals(web, resource)
    }
}
